docs/proposals/cluster-deployment.md

```diff
@@ -0,0 +1,190 @@
+
+**Substeps**:
+
+1. copy certificates
+2. copy manifest with:
+	1. etcd
+	2. apiserver
+	3. controller manager
+	4. scheduler
+3. run kubelet in docker container (configuration is read from apiserver Config object)
+4. run kubelet-checker in docker container
+
+**v1.2 simplifications**:
+
+1. kubelet-runner.sh - we will provide a custom docker image to run kubelet; it will contain
+kubelet binary and will run it using ```nsenter``` to workaround problem with mount propagation
+1. kubelet config file - we will read kubelet configuration file from disk instead of apiserver; it will
+be generated locally and copied to all nodes.
+
+**Exit criteria**:
+
+1. Can run basic API calls (e.g. create, list and delete pods) from the client side.
+
+### Step 4: Deploy nodes
+
+**Objective**: Run kubelet on all nodes.
+
+**Substeps**:
+
+1. copy certificates
+2. run kubelet in docker container (configuration is read from apiserver Config object)
+3. run kubelet-checker in docker container
+
+**v1.2 simplifications**:
+
+1. kubelet config file - we will read kubelet configuration file from disk instead of apiserver; it will
+be generated locally and copied to all nodes.
+
+**Exit critera**:
+
+1. All nodes are registered and ready.
+2. Can create a pod, which will be scheduled and run
+
+### Step 5: Setup kubernetes networking
+
+**Objective**: Configure the Kubernetes networking to allow routing requests to pods and services.
+
+To keep default setup consistent across open source deployments we will use Flannel to configure
+kubernetes networking. However, implementation of this step will allow to easily plug in different
+network solutions.
+
+**Substeps**:
+
+1. copy manifest for flannel server to master machine
+2. create a daemonset with flannel daemon (it will read assigned CIDR and configure network appropriately).
+
+**v1.2 simplifications**:
+
+1. flannel daemon will run as a standalone binary (not in docker container)
+2. flannel server will assign CIDRs to nodes outside of kubernetes; this will require restarting kubelet
+after reconfiguring network bridge on local machine; this will also require running master nad node differently
+(```--configure-cbr0=false``` on node and ```--allocate-node-cidrs=false``` on master), which breaks encapsulation
+between nodes
+
+**Exit criteria**:
+
+1. Can create a pod which will be scheduled and run and is accessible from all nodes
+2. Can create a service which is accessible from all nodes.
+
+### Step 6: Add daemons
+
+**Objective:** Start all system daemons (e.g. kube-proxy)
+
+**Substeps:**:
+
+1. Create daemonset for kube-proxy
+
+### Step 7: Add add-ons
+
+**Objective**: Add default add-ons (e.g. dns, logging)
+
+**Substeps:**:
+
+1. Create Deployments (and daemonsets if needed) for all add-ons
+
+## Deployment technology
+
+We will use Ansible as the default technology for deployment orchestration. It has low requirements on the cluster machines
```

Comment:
We had a long discussion with @brendandburns, @roberthbailey and @mikedanese about technology and what we really care about. I'll try summarize our reasoning here.

First of all we don't really care about the technology itself as long as it provides a way to use MIG-like deployments (both Ansible and cloud-init meet this requirement). What is important however is that we structure scripts/deployment in a way that each node can be deployed separately from other nodes. This will allow us to avoid cross-node dependencies and will make it easy to add additional nodes. We hope that structuring script properly will help us to avoid breaking theses assumptions in the future.

As it seems that Ansible is more adopted among our community we suggest to use it as the technology to implement deployment machinery.

@brendandburns, @roberthbailey @mikedanese - Did I miss something?